File: BatallonPython/peloton.py
from __future__ import annotations
from typing import List
from enums import TipoPeloton
from escuadra import Escuadra


class Peloton:
    """
    Clase Peloton
    Atributos:
      - nombre: str
      - comandante: str
      - tipo: TipoPeloton
      - escuadras: List[Escuadra]
      - max_escuadras: int (class attribute)
    """

    max_escuadras: int = 5

    def __init__(self, nombre: str, comandante: str, tipo: TipoPeloton):
        if not isinstance(tipo, TipoPeloton):
            raise TypeError("tipo debe ser TipoPeloton")
        self.nombre: str = nombre
        self.comandante: str = comandante
        self.tipo: TipoPeloton = tipo
        self.escuadras: List[Escuadra] = []

    def agregar_escuadra(self, e: Escuadra) -> bool:
        if e is None:
            raise ValueError("escuadra no puede ser None")
        if len(self.escuadras) >= Peloton.max_escuadras:
            return False
        self.escuadras.append(e)
        return True

    def remover_escuadra_por_nombre(self, nombre: str) -> bool:
        nombre_lower = nombre.lower()
        for i, e in enumerate(self.escuadras):
            if e.nombre.lower() == nombre_lower:
                del self.escuadras[i]
                return True
        return False

    def contar_escuadras(self) -> int:
        return len(self.escuadras)

    def contar_total_soldados(self) -> int:
        return sum(e.contar_soldados() for e in self.escuadras)

    def es_peloton_completo(self) -> bool:
        return len(self.escuadras) >= Peloton.max_escuadras

    @classmethod
    def cambiar_max_escuadras(cls, nuevo_max: int) -> None:
        if nuevo_max < 0:
            raise ValueError("nuevo_max debe ser >= 0")
        cls.max_escuadras = nuevo_max

    def descripcion(self) -> str:
        return (f"Pelotón {self.nombre} | Comandante: {self.comandante} | Tipo: {self.tipo.name} | "
                f"Escuadras: {len(self.escuadras)} | Soldados totales: {self.contar_total_soldados()}")

File: BatallonCsharp/Models/Peloton.cs
using Enums;

namespace Models
{
    public class Peloton
    {
        private string nombre;
        private string comandante;
        private TipoPeloton tipo;
        private List<Escuadra> escuadras;

        private static int maxEscuadras = 5;

        public Peloton(string nombre, string comandante, TipoPeloton tipo)
        {
            this.nombre = nombre;
            this.comandante = comandante;
            this.tipo = tipo;
            escuadras = new List<Escuadra>();
        }

        public bool AgregarEscuadra(Escuadra e)
        {
            if (escuadras.Count < maxEscuadras)
            {
                escuadras.Add(e);
                return true;
            }
            return false;
        }

        public bool RemoverEscuadraPorNombre(string nombre)
        {
            var escuadra = escuadras.FirstOrDefault(e => e.Nombre.Equals(nombre, StringComparison.OrdinalIgnoreCase));
            if (escuadra != null)
            {
                escuadras.Remove(escuadra);
                return true;
            }
            return false;
        }

        public int ContarEscuadras()
        {
            return escuadras.Count;
        }

        public int ContarTotalSoldados()
        {
            return escuadras.Sum(e => e.ContarSoldados());
        }

        public bool EsPelotonCompleto()
        {
            return escuadras.Count == maxEscuadras;
        }

        public static void CambiarMaxEscuadras(int nuevoMax)
        {
            maxEscuadras = nuevoMax;
        }

        public string Nombre => nombre;
        public string Comandante => comandante;
        public TipoPeloton Tipo => tipo;
        public List<Escuadra> Escuadras => escuadras;
    }
}
